app/services/avatar_generator.rb

```diff
@@ -0,0 +1,51 @@
+class AvatarGenerator
+  MAX_SIZE = 5
+  CONTENT_TYPES = %w[image/jpeg image/pjpeg image/png image/x-png]
+
+  attr_accessor :errors
+
+  def initialize(user, params)
+    @user   = user
+    @params = params
+    @avatar = params[:avatar]
+    @errors = []
+
+    validate_avatar
+  end
+
+  def call
+    @user.avatar.purge if @user.avatar.attached?
+
+    @user.avatar.attach(
+      io: process_image,
+      filename: @user.username,
+      content_type: @avatar.content_type
+    )
+  end
+
+  private
+
+  def process_image
+    orig_width = @params[:original_width].to_i
+    width      = @params[:height_width].to_i
+    left       = @params[:width_offset].to_i
+    top        = @params[:height_offset].to_i
+
+    ImageProcessing::MiniMagick.
+      source(@avatar.tempfile).
+      resize_to_fit(orig_width, nil).
+      crop("#{width}x#{width}+#{left}+#{top}!").
+      convert("png").
+      call
```

Comment:
won't this be an expensive operation risking a time out? it should perhaps be moved to a background job in a separate PR.